### ISSUE TITLE
Provide ability to show screenshots in search results

### ISSUE THREAD

**Issue by StefanSchroeder** [NixOS/nixos-search#211]
This is an improvement idea.

If a meta-attribute "Screenshot-URL" were added to the package description, the search result could include a screenshot of the application in the result. There is no need to create new screenshots as http://screenshots.debian.net/ provides a public repository of screenshots already that could be reused.

I am aware that adding screenshots to the existing thousands of packages would be a major undertaking, but I think it would be a nice user experience enhancement and - of course - optional, since screenshots only make sense for graphical applications.

When the Screenshot-URL is provided for a package, the search result could fetch the image and show it alongside the package description.

**Comment by garbas** [comment]
Idea is great. I had something similar in mind as we could also show the icon.

That said there are few problems with current proposal. Nix package names are not the same as Debian package names.

It would be not a problem to add images in the search frontend, but we would need to standardize this first in nixpkgs. I think first step would be to open a PR with a proposed change to the example package (eg. firefox). Then the discussion will happen, which will probably require us to document this and update instructions for package maintainers.

So in generally I'm :100: for this change, but work needs to happen in nixpkgs first.

I will keep this issue open but marked it as low priority.

**Comment by StefanSchroeder** [comment]
Adding the Screenshot-URL in the meta-attributes has no side-effects. There is no need to have matching nixos-package names and debian-package names, images could be hosted anywhere, even upstream. The URLs would have to be handpicked by the maintainers. The front-end could enforce an image size; the URL http://screenshots.debian.net/thumbnail/PACKAGENAME is small and could probably shown without restrictions.
Where would we start? First the PR for adding the URL to a package or first changing the manual, https://nixos.org/manual/nixpkgs/stable/#sec-standard-meta-attributes ?

**Comment by StefanSchroeder** [comment]
@garbas Would you be willing to mentor me through the process? I've never contributed to Nixos and frankly the manuals are not always precise or detailed enough for new contributors.

**Comment by garbas** [comment]
@StefanSchroeder Of course, I might be unresponsive from time to time due to workload, but lets try to get this into nixpkgs.

You already found the place where we would need to document it. Now lets add new meta attribute. Here are my thoughts, what should be in the PR:

- I wouldn't limit our self to only screenshorts.debian.net but allow users to specify any url. But I would mention that we will took inspiration from screenshots.debian.net.

- I would provide an example, firefox.

- I would propose we add two attributes, one for thumbnail (160x120 pixels or less) and one for screenshot (800x600 pixels or less). I took the dimensions from http://screenshots.debian.net/about. We could either propose:

```nix
meta.screenshotThumb = "https://...";
meta.screenshotLarge = "https://...";
```

or

```nix
meta.screenshot =
  { thumb = "https://...";
    large = "https://...";
  };
```

- I would also provide link to this discussion.

Would this help you move forward?

**Comment by StefanSchroeder** [comment]
I have now made two changes in my fork: One in the Nixpkg-manual to document the new meta-attribute. One to apply the new attributes to the "pingus" package. 

Open questions: 
- I haven't created a feature-branch in my repo. Is that required? 
- I haven't built the manual to check for syntax errors in the XML. I have to yet figure out how to do that. Should be doable.
- I am not sure if the Nix package manager will actually permit the new attributes or if Nix will decline to process the pingus/default.nix due to an unknown attribute. (In other words, I don't know if new attributes must be added to Nix)

See here:
https://github.com/NixOS/nixpkgs/compare/master...StefanSchroeder:master

**Comment by garbas** [comment]
> I have now made two changes in my fork: One in the Nixpkg-manual to document the new meta-attribute. One to apply the new attributes to the "pingus" package.
> 
> Open questions:
> 
>     * I haven't created a feature-branch in my repo. Is that required?

This all depends how you want to work. If you will be working on two changes you will need to use branches.

>     * I haven't built the manual to check for syntax errors in the XML. I have to yet figure out how to do that. Should be doable.

```console
nix-build nixos/release.nix -A manualHTML.x86_64-linux
```

The documentation should be in `./result/.

>     * I am not sure if the Nix package manager will actually permit the new attributes or if Nix will decline to process the pingus/default.nix due to an unknown attribute. (In other words, I don't know if new attributes must be added to Nix)

Nix does not check any of these values. No change in Nix is needed.

> See here:
> [NixOS/nixpkgs@master...StefanSchroeder:master](https://github.com/NixOS/nixpkgs/compare/master...StefanSchroeder:master)

I would also mention:
- where one can find screenshots
- what is the expected size of the images
- provide and example with `<programlisting> ... example goes here </programlisting>`

**Comment by StefanSchroeder** [comment]
Fixed as suggested.
https://github.com/NixOS/nixpkgs/compare/master...StefanSchroeder:master

**Comment by garbas** [comment]
@StefanSchroeder Great, now open pull request and explain reasons for this change. Then we should leave it for a week or more for people to comment on it and then see how to finalize this.

**Comment by StefanSchroeder** [comment]
Almost there. One final verification step is missing. The nix-build command that you provided builds the "Nixos" manual. The changes I supplied to document the meta-attributes are going into the "Nixpkgs" manual.  I was unable to figure out how to build the Nixpkgs manual.

**Comment by garbas** [comment]
@StefanSchroeder ah right, I always confuse them myself as well :) then command for nixpkgs is

```console
$ nix-build doc/default.nix
```

**Comment by StefanSchroeder** [comment]
@garbas I created the PR. Some automatic CICD pipeline was kicked off that performed a variety of checks.
One of the checks fails and I don't know what the problem is:
https://github.com/NixOS/nixpkgs/actions/runs/322959702 
The title of the failed check is "Checking EditorConfig". I am familiar with what EditorConfig does, but I get no indication, where the formatting slip occurs.

**Comment by garbas** [comment]
@StefanSchroeder you need to click deeper what it failed with 

here is the direct link: https://github.com/NixOS/nixpkgs/runs/1295098114?check_suite_focus=true
and here is the screenshot: ![2020-10-23-153344_1220x765_scrot](https://user-images.githubusercontent.com/20208/97010199-610b7680-1545-11eb-9068-e40782a6ced8.png)

I think you need to remove some whitespaces at the end of two lines.

**Comment by StefanSchroeder** [comment]
ah, ok.
